Source/DafnyStandardLibraries/src/Std/Actions/Aggregators.dfy

```diff
@@ -0,0 +1,299 @@
+/*******************************************************************************
+ *  Copyright by the contributors to the Dafny Project
+ *  SPDX-License-Identifier: MIT
+ *******************************************************************************/
+
+module Std.Aggregators {
+
+  import opened Actions
+  import opened Wrappers
+  import opened DynamicArray
+  import Collections.Seq
+
+  // TODO: Not 100% confident in the exact name and signature here:
+  // * As per the module name, I think I like "Aggregator" as the dual to "Enumerator" better
+  // * This signature is not a complete dual of Enumerator: an Enumerator is a finite enumeration,
+  //   and one where the number of elements isn't statically known
+  //   (returns an Option<T> rather than a T). A dual to that concept should also be allowed
+  //   to accept only a finite number of elements, so it should probably return a boolean.
+  //   This would be the kind of Action that writes element in sequence into a fixed-length array,
+  //   for example, and therefore has finite capacity.
+  //   
+  //   Perhaps we should have:
+  //     - IEnumerator = Action<(), T> (potentially infinite enumeration)
+  //     - Enumerator = Action<(), Option<T>> + proofs of eventually producing None (finite enumeration)
+  //     - IAggregator = Action<T, ()> (potentially infinite aggregation)
+  //     - Aggregator = Action<T, boolean> + proofs of eventually producing false (finite aggregation)
```

Comment:
I like the distinction between `IEnumerator` and `Enumerator` and see how that could be useful. But, your comments about duality notwithstanding, I don't see a practical use for a "can only accept a finite number of things" `Aggregator`.

Reply:
I think there are definitely cases of aggregators with "finite capacity". For example, an Aggregator that stores consumed values into an array in order. That could be either an `IAggregator` with a `CanConsume` precondition that the aggregator still has room, or an `Aggregator` that returns `false` if it's out of room.

I realize now that I have an `ArrayAggregator` that uses a `DynamicArray` and hence doesn't have a fixed capacity. I'll rename that to `DynamicArrayAggregator` and also include an `ArrayAggregator` to motivate this better.